### ISSUE TITLE
Wayland not always isntalled with wl-copy/wl-paste; clipboard stop working unexpectedly

### ISSUE THREAD

**Issue by mlittlej** [neovim/neovim#26196]
### Problem

Switched my computer (Debian (testing)) from using X11 to Wayland and shortly thereafter realised that I'd lost the ability to yank and paste to/from the * and + registers.

Attempts gave the following error:
```
1 line yanked into "*                                                                         
clipboard: error invoking xclip: Authorization required, but no authorization protocol specified  Error:
 Can't open display: :0 
clipboard: error: Authorization required, but no authorization protocol specified
```

Checked `:help clipboard` which says neovim determines how to use those registers by checking the the `$WAYLAND_DISPLAY` env var. Mine was set:

```
$ echo $WAYLAND_DISPLAY 
wayland-0
```

A bit of quick digging on StackOverflow later led me to the fact that neovim uses `wl-copy` and `wl-paste` to enable the expected behaviour, thereupon I found out that I don't have those commands installed. A quick `apt install wl-clipboard` later and the problem is solved.

I looked at the code that figures out how to set the clipboard [here][1] and it's not only checking for `$WAYLAND_DISPLAY`, it's also checking whether or not `wl-copy` and `wl-paste` are installed which is fair enough, but that isn't exactly what's documented in the `:help`.

A further read leads me to the condition I actually fell into (line 113) `elseif !empty($DISPLAY) && executable('xclip')`:
```
$ echo $DISPLAY 
:0
$ type xsel
bash: type: xsel: not found
$ type xclip
xclip is aliased to `xclip -selection clipboard'
```

... which explains why it was trying to call `xclip` and why I never got any error messages on startup as I might have expected.

I think there's two problem:

1. There's a documentation problem that the `:help` doesn't fully describe the checks that are actually done when figuring out which clipboard tool to use. I think this could be fixed fairly easily to give a more exact description.

2. The logic in `provider#clipboard#Executable()` with respect to Wayland, and perhaps others, isn't robust enough. I'm not _au fait_ with vimscript, but maybe something like the following would be better:
```
[...]
  elseif !empty($WAYLAND_DISPLAY) 
    if executable('wl-copy') && executable('wl-paste')
      let s:copy['+'] = ['wl-copy', '--type', 'text/plain']
      let s:paste['+'] = ['wl-paste', '--no-newline']
      let s:copy['*'] = ['wl-copy', '--primary', '--type', 'text/plain']
      let s:paste['*'] = ['wl-paste', '--no-newline', '--primary']
      return 'wl-copy'
    elseif executable('waycopy') && executable('waypaste')
      let s:copy['+'] = ['waycopy', '-t', 'text/plain']
      let s:paste['+'] = ['waypaste', '-t', 'text/plain']
      let s:copy['*'] = s:copy['+']
      let s:paste['*'] = s:paste['+']
      return 'wayclip'
    else
      echohl WarningMsg
      echomsg 'clipboard: error detecting clipboard tool. Wayland session detected, no wayland clipboard tools installed'
      echohl None
      return ''
[...]
```

[1]: https://github.com/neovim/neovim/blob/53cc835b51546f9ff68facebab2e0c6e97585e5b/runtime/autoload/provider/clipboard.vim#L95

### Steps to reproduce

* Run under a Wayland Windows manager session ($WAYLAND_DISPLAY envvar should be non-empty)
* Don't have the `wl-copy` and `wl-paste`, or `waycopy` and `waypaste` tools installed
* Also have the $DISPLAY envvar non-empty
* Have `xclip` installed (`xsel` might do funny things as well, no idea)
* Try to yank anything into the * or + registers

### Expected behavior

Neovim should correctly detect the error condition of running a Wayland session without the correct Wayland Clipboard tools installed and notify the user of the error.

It should not be able to fall through to another apparently correct set of clipboard boards which do not work.

The documentation should be more exact in describing how the clipboard functionality is determined to enable a faster user self-diagnostic and repair.

### Neovim version (nvim -v)

NVIM v0.7.2

### Vim (not Nvim) behaves the same?

maybe, didn't fully test but on yank it does emit a similar error message. VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Sep 11 2023 13:22:50)

### Operating system/version

Debian trixie/sid

### Terminal name/version

gnome-terminal

### $TERM environment variable

xterm-256color

### Installation

apt

**Comment by clason** [comment]
0.7.2 is ancient. Please always test with current releases (0.9.4 at time of writing).

**Comment by mlittlej** [comment]
0.7.2 is what I have installed, it's where the issue was observed and it's what's available via `apt` in Debian trixie/sid, hence it's the version number I'm reporting the issue under.

I could install 0.9.4 and uninstall `wl-clipboard`, but I suspect I'll get the same results. I've looked at the release-0.7 branch code and the logic has the same flaw as the latest code of master - the older version doesn't check for the `waycopy` and `waypaste` tools.

The lack of detection of the error condition and the lack of reporting of that error condition remain an issue - even in the latest version. I'd be obliged if you'd at least re-open the ticket on that basis.

I accept that the documentation is correct, I've misread/misunderstood it.

**Comment by clason** [comment]
Nevertheless, 0.7.x is no longer supported, so any issues there are irrelevant unless confirmed with the latest version. 

> but I suspect I'll get the same results.

In general, don't guess; verify. But yes, this is still the documented behavior. And have you looked at `:checkhealth`?

**Comment by mlittlej** [comment]
This is from the 0.9.4 AppImage (installed per [here][1]), with the `wl-clipboard` apt package uninstalled.

`:checkhealth` (which I had forgotten about, thank you for reminding me)

```
    28 provider: health#provider#check
    29 
    30 Clipboard (optional)
    31 - OK Clipboard tool found: xclip
    32 
```

NB: Doing this in 0.9.4 highlighted the issue via a different mechanism. The default behaviour of 0.9.4 seems to set `mouse=nvi` which captured my mouse when I highlighted the above text, and on trying to copy with a right click, failed with the same error message: 
```
clipboard: error invoking xclip: Authorization required, but no authorization protocol specified  Error:
 Can't open display: :0 
```

The copy option was presented via a neovim popup window/dialogue (apols. I don't know the correct term for this UI element, but it's a neovim thing) rather than the gnome right-click menu I'd expected. I had to set `mouse=` to enable my "normal" gnome highlighting and right-click menu to be able to copy the above text out.

But FWIW the issue is still present in v0.9.4 in in it's originally reported `"*y` form as well.

---

After installing `wl-clipboard` again:

```
provider: health#provider#check

Clipboard (optional) ~
- OK Clipboard tool found: wl-copy
```
NB: That was using the mouse without fiddling with `:set mouse`, hence it doesn't include the number line.

And the `"*y` was working again as well.

[1]: https://github.com/neovim/neovim/wiki/Installing-Neovim#appimage-universal-linux-package

**Comment by clason** [comment]
See? So documentation and reporting is working and helpful.

**Comment by mlittlej** [comment]
I agreed previously that the documentation was correct and that I'd misunderstood it. I hadn't used `:checkhealth` reporting but have reason to believe it's been anything but working.

But I still believe that code that allows `xclip` to be selected as the clipboard utility, while neovim was running under a Wayland WM Session where `xclip` cannot and does not work, which in turn breaks neovim yank-to-selection-register functionality, has an error in it.

Do you disagree?

**Comment by clason** [comment]
I disagree. Platform detection from inside a cross-platform terminal program is _hard_, and we have to stick to "best effort" -- plus offering the possibility of overriding, which we do.
